katdal/dataset.py

```diff
@@ -227,29 +227,43 @@ def _calc_target_coords(cache, name, ant, projection, coordsys):
     return x if name.startswith(ant_group + 'target_x') else y
 
 
-def _calc_uvw(cache, name, antA, antB):
-    """Calculate (u,v,w) coordinates using sensor cache contents."""
-    antennaA = cache.get('Antennas/%s/antenna' % (antA,))[0]
-    antennaB = cache.get('Antennas/%s/antenna' % (antB,))[0]
-    u = np.empty(len(cache.timestamps))
-    v = np.empty(len(cache.timestamps))
-    w = np.empty(len(cache.timestamps))
+def _calc_uvw_basis(cache, name, ant):
+    """Calculate (u,v,w) basis vectors using sensor cache contents."""
+    ant_group = 'Antennas/%s/' % (ant,)
+    antenna = cache.get(ant_group + 'antenna')[0]
+    u = np.empty((len(cache.timestamps), 3))
+    v = np.empty((len(cache.timestamps), 3))
+    w = np.empty((len(cache.timestamps), 3))
     targets = cache.get('Observation/target')
     for segm, target in targets.segments():
-        u[segm], v[segm], w[segm] = target.uvw(antennaA, cache.timestamps[segm],
-                                               antennaB)
-    cache['Corrprods/%s/%s/u' % (antA, antB)] = u
-    cache['Corrprods/%s/%s/v' % (antA, antB)] = v
-    cache['Corrprods/%s/%s/w' % (antA, antB)] = w
-    return u if name.endswith('u') else v if name.endswith('v') else w
+        basis = target.uvw_basis(cache.timestamps[segm], antenna)
+        u[segm], v[segm], w[segm] = basis.transpose(0, 2, 1)
```

Comment:
Maybe worth indicating which axes you're swapping in a comment.

Also, is it necessary to permute here, rather than feeding a different axis to tensordot below?

Reply:
Added comments to explain the reshaping. It is necessary to permute, since a sensor expressed as an array must have time as the first dimension. On the other hand, I've now realised that the cargo-copied tensordot is overkill - a simple `dot` will do since `basis` is a T x 3 array.